docs/operations/loki-canary.md

```diff
@@ -0,0 +1,172 @@
+# Loki Canary
+
+Loki Canary is a standalone app that audits the log capturing performance of
+Loki.
+
+## How it works
+
+![block_diagram](loki-canary-block.png)
+
+Loki Canary writes a log to a file and stores the timestamp in an internal
+array. The contents look something like this:
+
+```nohighlight
+1557935669096040040 ppppppppppppppppppppppppppppppppppppppppppppppppppppppppppp
+```
+
+The relevant part of the log entry is the timestamp; the `p`s are just filler
+bytes to make the size of the log configurable.
+
+An agent (like Promtail) should be configured to read the log file and ship it
+to Loki.
+
+Meanwhile, Loki Canary will open a WebSocket connection to Loki and will tail
+the logs it creates. When a log is received on the WebSocket, the timestamp
+in the log message is compared to the internal array.
+
+If the received log is:
+
+* The next in the array to be received, it is removed from the array and the
+  (current time - log timestamp) is recorded in the `response_latency`
+  histogram. This is the expected behavior for well behaving logs.
+* Not the next in the array received, is is removed from the array, the
+  response time is recorded in the `response_latency` histogram, and the
+  `out_of_order_entries` counter is incremented.
+* Not in the array at all, it is checked against a separate list of received
+  logs to either increment the `duplicate_entries` counter or the
+  `unexpected_entries` counter.
+
+In the background, Loki Canary also runs a timer which iterates through all of
+the entries in the internal array. If any of the entries are older than the
+duration specified by the `-wait` flag (defaulting to 60s), they are removed
+from the array and the `websocket_missing_entries` counter is incremented. An
+additional query is then made directly to Loki for any missing entries to
+determine if they are truly missing or only missing from the WebSocket. If
+missing entries are not found in the direct query, the `missing_entries` counter
+is incremented.
+
+## Installation
+
+### Binary
+
+Loki Canary is provided as a pre-compiled binary as part of the
+[Loki Releases](https://github.com/grafana/loki/releases) on GitHub.
+
+### Docker
+
+Loki Canary is also provided as a Docker container image:
+
+```bash
+# change tag to the most recent release
+$ docker pull grafana/loki-canary:v0.2.0
+```
+
+### Kubernetes
+
+To run on Kubernetes, you can do something simple like:
+
+`kubectl run loki-canary --generator=run-pod/v1
+--image=grafana/loki-canary:latest --restart=Never --image-pull-policy=Never
+--labels=name=loki-canary -- -addr=loki:3100`
+
+Or you can do something more complex like deploy it as a DaemonSet, there is a
```

Comment:
@slim-bean Could you provide more detail on this?

Reply:
Okay, I have an answer to this now :slightly_smiling_face: Since we also recommend running promtail as a DaemonSet, I think it makes sense for Loki Canary to be a DaemonSet too; the point is for it to have a replica running on each node, and doing it like that while you have Promtail helps you audit the nodes that Promtail is operating on.